### ISSUE TITLE
Make (TLS) groups fully independent of (EC) curves

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#13767]
Referring to the statement by @mattcaswell in #10512

> We have implemented fully pluggable KEM and KEX in libssl

wouldn't the expectation be fair that after adding new TLS (1.3) groups (via a pluggable KEM provider) those new groups should be visible/accessible in a similar way as the original (EC) curves? 

More specifically, the API `SSL_get_shared_group` seems to behave somewhat out of line with this full integration: By returning only NIDs (of known libcrypto implementations), it effectively returns nothing if a [new, Provider-implemented, TLS1.3 KEM group (PQC in this case)](https://github.com/open-quantum-safe/openssl/tree/OQS-OpenSSL3) has been integrated and runs:

```
Supported Elliptic Groups: 0x0201
Shared Elliptic groups: <NULL>
```

TLS1.3 operates correctly with the new crypto (KEM) from the Provider (confirming the statement at the beginning), but only tries to display built-in/libcrypto-provided shared (Elliptic) groups.

In any case, the output above seems to be inconsistent: Either it should display nothing (as the new KEM is not an EC curve) OR also display the shared group (which already is shown as supported with its group ID). The reason for the NULL above is this statement in `ssl/s3_lib.c`:

```
return tls1_group_id2nid(id, 1);
```

A possible alternative is to amend the above return statement as follows

```
if (s->session->ssl_version != TLS1_3_VERSION) return tls1_group_id2nid(id, 1);
```

and update the documentation in `doc/man3/SSL_CTX_set1_curves.pod` accordingly: Amend the current

> SSL_get_shared_group() returns the NID of the shared group B<n> for a server-side SSL B<ssl>.

as follows:

> SSL_get_shared_group() returns the NID of the shared group B<n> for a server-side SSL B<ssl> when running SSL or TLS<1.3; for TLS1.3, returns the shared group ID.

In addition, it would be nice to amend the code in `apps/lib/s_cb.c` annotated "TODO(TLS1.3)" to fetch the group name. When trying to implement this I had to find I'm out of my depth here: At first look, the function `tls1_group_id_lookup` could deliver the `TLS_GROUP_INFO` structure containing all required information, however, function and data structures seem to be only (SSL-)local and not generally accessible, so some internal "API opening" may be be required for this. So, any reference to APIs I am not aware of and already delivering this functionality would be gratefully welcomed: I'd volunteer to integrate that into `apps/lib/s_cb.c`.

An alternative to this proposal would be a new API for TLS1.3 groups that has no relation to the old (EC) curves.

**Comment by romen** [comment]
Thanks for reporting this (and the testing of pluggable KEM groups backed by liboqs implementations).

I think both solutions would be acceptable and it is great you are willing to work on this.

I'd defer to @mattcaswell for more direction on which approach would be preferable.

**Comment by romen** [comment]
I'm adding this to the 3.0beta1 milestone for evaluation by OTC: while the addition of a new API or opening existing ones could configure as a new feature, the issue highlights a shortcoming in the current `SSL_get_shared_group()` implementation which should be discussed.

**Comment by mattcaswell** [comment]
I think the problem is purely in the implementation of `ssl_print_groups` in apps\s_cb.c.

`SSL_get_shared_group()` is documented to behave as follows in the event of a non-built-in group:

> If the NID for the shared group is unknown then the value
is set to the bitwise OR of TLSEXT_nid_unknown (0x1000000) and the id of the
group.

This behaviour should already be implemented in `tls1_group_id2nid`, so `SSL_get_shared_group` should already be doing the right thing.

The "Supported Elliptic Groups" output already has the logic to handle this:

https://github.com/openssl/openssl/blob/2c61a670ebf2f1923a3bd2ef0ee4b2fa6261eaeb/apps/lib/s_cb.c#L356-L371

But the "Shared Elliptic Groups" output is missing this logic:

https://github.com/openssl/openssl/blob/2c61a670ebf2f1923a3bd2ef0ee4b2fa6261eaeb/apps/lib/s_cb.c#L378-L389

We just need to add the missing logic and it should work as expected.

While we're at it, I'd suggest we drop the word "Elliptic" from the labels and just make them "Supported Groups" and "Shared Groups".

> In addition, it would be nice to amend the code in apps/lib/s_cb.c annotated "TODO(TLS1.3)" to fetch the group name. When trying to implement this I had to find I'm out of my depth here: At first look, the function tls1_group_id_lookup could deliver the TLS_GROUP_INFO structure containing all required information, however, function and data structures seem to be only (SSL-)local and not generally accessible, so some internal "API opening" may be be required for this. So, any reference to APIs I am not aware of and already delivering this functionality would be gratefully welcomed

I don't think there is an API to do this at the moment. We'd have to add something new. Say, something like:

````
    const char *SSL_group_to_name(SSL *s, int group);
````
This could simply take the id in the format provided by `SSL_get_shared_group` and return the associated string value for it.

**Comment by baentsch** [comment]
Thanks for your quick feedback & Thanks for the pointer to the `TLSEXT_nid_unknown` logic. [Here's a fix](https://github.com/openssl/openssl/compare/master...baentsch:issue-13767?expand=1) along your suggestions, @mattcaswell.
As I don't know your team's prerequisites for PRs (beyond running `make test` with 100% PASS), I refrained from just quickly submitting the above to close out the immediate bug. I'd also be glad to add the "groupname-getter" API in the PR if you'd be accepting such--clearly non-essential--code before getting the beta out the door. Glad to take your guidance as to how/whether/when to proceed.

**Comment by mattcaswell** [comment]
> Thanks for your quick feedback & Thanks for the pointer to the TLSEXT_nid_unknown logic. Here's a fix along your suggestions, @mattcaswell.

The fix looks good from my quick scan. Please feel free to go ahead and create a PR for it. The main thing you may need to do is send us a CLA (if you haven't already done so):

https://www.openssl.org/policies/cla.html

> I'd also be glad to add the "groupname-getter" API in the PR if you'd be accepting such--clearly non-essential--code before getting the beta out the door. Glad to take your guidance as to how/whether/when to proceed.

We haven't yet reached feature freeze, so we still have the ability to add new APIs if we feel it is warranted. In this case the API should be relatively straightforward and low risk - and the need for it is introduced by other changes already made for 3.0. So, I don't see a problem with us doing this now - although its unlikely that we would hold up 3.0 beta1 for it, and if it missed beta1 it wouldn't go in for somewhile. However, there's still a lot to do before beta1 so this shouldn't be an issue.

So, I'd be happy to see you submit a PR to add this. It would need to be accompanied with appropriate documentation and a test (probably in test/sslapitest.c, perhaps as an extension of the existing `test_pluggable_group` test).

**Comment by baentsch** [comment]
Thanks again for feedback & guidance.

> The main thing you may need to do is send us a CLA (if you haven't already done so)

So done now. Will I get an ACK on this or can I just proceed with the PR?

> you submit a PR to add this. It would need to be accompanied with appropriate documentation and a test (probably in test/sslapitest.c, perhaps as an extension of the existing test_pluggable_group test).

OK - working on it....

Additional, immediate questions after checking out the "surrounding" code for this addition: Should I make "related" fix suggestions in the same PR or do separate issue/PR pairs on things that don't look right? Most notably, I spotted many #ifdef's that don't look quite right any more: For example, the ones below appear out of step with the new groups' stance as EC-independent entities now (and there are some more). 

Key issue: The internal s3 data structure headers (correctly) don't contain #ifdef's any more for elements like [`peer_tmp`](https://github.com/openssl/openssl/blob/2c61a670ebf2f1923a3bd2ef0ee4b2fa6261eaeb/ssl/ssl_local.h#L1420) or [`pkey`](https://github.com/openssl/openssl/blob/2c61a670ebf2f1923a3bd2ef0ee4b2fa6261eaeb/ssl/ssl_local.h#L1306) but code like this doesn't handle them right if EC and DH are both undefined:

https://github.com/openssl/openssl/blob/2c61a670ebf2f1923a3bd2ef0ee4b2fa6261eaeb/ssl/s3_lib.c#L3398-L3401

Also the command dispatchers seem to be flawed in this case:

https://github.com/openssl/openssl/blob/2c61a670ebf2f1923a3bd2ef0ee4b2fa6261eaeb/ssl/s3_lib.c#L3625-L3671

Additional proof point: when I build my (previously perfectly functioning) PQC-providers in an OpenSSL master branch set up with `./Configure -no-ec -no-dh` nothing works any more. Reason is clear: The groups' `SSL_CTRL_XXX` commands simply are not present/executed due to code missing when `-no-ec -no-dh` (OPENSSL_NO_EC, OPENSSL_NO_DH) are set. 

Looking at all of this I guess it's reasonable to just make change suggestions in the same PR as it's all got to do with this issue -- and it's already labelled "bug" :)

**Comment by kaduk** [comment]
> Additional, immediate questions after checking out the "surrounding" code for this addition: Should I make "related" fix suggestions in the same PR or do separate issue/PR pairs on things that don't look right? Most notably, I spotted many #ifdef's that don't look quite right any more: For example, the ones below appear out of step with the new groups' stance as EC-independent entities now (and there are some more).

Related fixes in the same PR should be fine.  Put them in separate commits, though, please.

**Comment by baentsch** [comment]
> Related fixes in the same PR should be fine. Put them in separate commits, though, please.

Thanks, understood: Makes things easier to understand/pick apart.

Now, unfortunately, prior to this suggestion I already did a one-commit PR, so I wonder whether all comments now received warrant withdrawing it and re-doing things bit-by-bit (fixing the original `apps` output error; then creating tests catching the #ifdef errors; fixing those errors; adding group-name retrieval function; adding tests for that (in a `-no-ec -no-dh` configured build); each step run through `make test` and accompanied by separate commit messages): Surely doable. It will create more work on all sides, incl. reviewing code that is changed at least twice in such process, though. So I wonder whether this strikes a good balance for what seem to me to be rather straightforward changes. But then again, this is security software and everyone should understand at first glance what's going on.

In sum: I'm happy to proceed either way: I don't want to break norms that have worked well for this project and ensure its quality. I shall rely on a review decision in the PR to guide me how to proceed with this. Either way, I'll surely apply lessons learned this time around for possible future contributions.

**Comment by mattcaswell** [comment]
I much prefer having separate commits for different items.

**Comment by baentsch** [comment]
> I much prefer having separate commits for different items.

Working on that (incl. incorporating all feedback). Part 1 is done: Only adding `SSL_group_to_name`.

Now looking at correcting the (so far presumed incorrect) behaviour of OpenSSL group functions and data structures if `-no-ec -no-dh` is set... When trying to activate a test using one of the then unavailable group functions, I noticed this:
https://github.com/openssl/openssl/blob/e260bee0a97d4e6de60eae2c86d7c11ed03f2010/Configure#L566-L567

If I get it right, this means that it is entirely intentional that TLS1_3 is not meant to work (at least `OPENSSL_NO_TLS1_3` seems to get auto-set in that case) if neither EC nor DH is present. Can you confirm this (is intentional)? 

A quick test after building openssl with EC and DH disabled (`./Configure -no-ec -no-dh`) seems to confirm this:

```
> ./apps/openssl s_server -cert rsa_srv.crt -key rsa_srv.key -www -tls1_3
s_server: Unknown option: -tls1_3
```

In addition, all specific TLS1_3 tests get disabled in `test` (incl. those for TLS groups).

If so, we'd be basically done with the PR as it stands as the combination no-ec&&no-dh&tls1_3(i.e., TLS group support) effectively is impossible.

**Comment by mattcaswell** [comment]
> If I get it right, this means that it is entirely intentional that TLS1_3 is not meant to work (at least OPENSSL_NO_TLS1_3 seems to get auto-set in that case) if neither EC nor DH is present. Can you confirm this (is intentional)?

It is intentional in the sense that this code was introduced at a time prior to the introduction of pluggable groups. At that point if you didn't have ec or dh then mostly TLSv1.3 couldn't work (although I suppose possibly it could be made to work if you used pre shared psks...). Now that we have pluggable groups I'm not sure this is still correct.

I have a feeling though that removing the no-ec/no-dh restrictions might turn into a much bigger job then you had intended (I'm wondering what the consequences throughout the test code might be). Since its a bit of scope creep from the original intent, you might want to consider dropping that part (or splitting it out into a separate PR so as not to derail the main purpose of your PR).

**Comment by romen** [comment]
> > If I get it right, this means that it is entirely intentional that TLS1_3 is not meant to work (at least OPENSSL_NO_TLS1_3 seems to get auto-set in that case) if neither EC nor DH is present. Can you confirm this (is intentional)?
> 
> It is intentional in the sense that this code was introduced at a time prior to the introduction of pluggable groups. At that point if you didn't have ec or dh then mostly TLSv1.3 couldn't work (although I suppose possibly it could be made to work if you used pre shared psks...). Now that we have pluggable groups I'm not sure this is still correct.
> 
> I have a feeling though that removing the no-ec/no-dh restrictions might turn into a much bigger job then you had intended (I'm wondering what the consequences throughout the test code might be). Since its a bit of scope creep from the original intent, you might want to consider dropping that part (or splitting it out into a separate PR so as not to derail the main purpose of your PR).

I completely agree, and I would also suggest to tackle the topic of revisiting the `no-ec+no-dc = no-tls` restrictions as a separate PR!

**Comment by romen** [comment]
> > > If I get it right, this means that it is entirely intentional that TLS1_3 is not meant to work (at least OPENSSL_NO_TLS1_3 seems to get auto-set in that case) if neither EC nor DH is present. Can you confirm this (is intentional)?
> > 
> > 
> > It is intentional in the sense that this code was introduced at a time prior to the introduction of pluggable groups. At that point if you didn't have ec or dh then mostly TLSv1.3 couldn't work (although I suppose possibly it could be made to work if you used pre shared psks...). Now that we have pluggable groups I'm not sure this is still correct.
> > I have a feeling though that removing the no-ec/no-dh restrictions might turn into a much bigger job then you had intended (I'm wondering what the consequences throughout the test code might be). Since its a bit of scope creep from the original intent, you might want to consider dropping that part (or splitting it out into a separate PR so as not to derail the main purpose of your PR).
> 
> I completely agree, and I would also suggest to tackle the topic of revisiting the `no-ec+no-dh = no-tls` restrictions as a separate PR!

In that PR, one test that would be nice to have is to make sure that in `no-ec+no-dh` builds, it is still possible to use the `xor-kex/xor-kem` groups plugged by the test provider!

**Comment by baentsch** [comment]
> In that PR, one test that would be nice to have is to make sure that in no-ec+no-dh builds, it is still possible to use the xor-kex/xor-kem groups plugged by the test provider!

That was exactly the test I wanted to activate when I "discovered" the !EC+!DH=!TLS1_3 "equation" and the consequences dawned an me (as this being more than what I originally signed up for).

With your two votes, I'd then also consider the PR feature complete from my end now.

**Comment by kaduk** [comment]
With #13785 merged, should this issue get closed as well?  Or is there more left to do?

**Comment by baentsch** [comment]
There is the problem remaining that OpenSSL doesn't properly support (and test) TLS1.3 if neither EC nor DH are configured (`./Configure -no-ec -no-dh`).

**Comment by richsalz** [comment]
If you disable both EC and DH, then there are no TLS 1.3 key exchange mechanisms left so TLS 1.3 should be disabled.

**Comment by beldmit** [comment]
In case of using the pluggable KEM it is still possibe, I think

**Comment by richsalz** [comment]
> In case of using the pluggable KEM it is still possibe, I think

ah right, thanks @beldmit

**Comment by baentsch** [comment]
Update after some code changes: It seems possible to break the (now outdated) link "no-ec&&no-dh -> no tls1_3" by adding `|| !defined(OPENSSL_NO_TLS1_3)` to places like https://github.com/openssl/openssl/blob/e211d949cd5737e53cd3399e6a88453930768b98/ssl/s3_lib.c#L3625-L3626 
to enable the TLS1.3 groups logic supported by external providers.

In consequence, though, all tests with the "old assumption" (i.e., no EC&&no DH means no available groups) fail. Again, they can be made to work by adding a suitable provider, e.g., like this:
```
#if defined(OPENSSL_NO_EC) && defined(OPENSSL_NO_DH) && \
    !defined(OPENSSL_NO_TLS1_3)

    OSSL_PROVIDER *tlsprov = OSSL_PROVIDER_load(libctx, "tls-provider");

    if (!TEST_ptr(tlsprov))
        goto end;
#endif
```
followed before connection establishment by setting the group to use:
```
#if defined(OPENSSL_NO_EC) && defined(OPENSSL_NO_DH) && \
    !defined(OPENSSL_NO_TLS1_3)

    if (!TEST_true(SSL_set1_groups_list(serverssl, "xorgroup"))
            || !TEST_true(SSL_set1_groups_list(clientssl, "xorgroup")))
        goto end;

#endif
```
This code would need to be added to all "old assumption" test cases. The second part doesn't only seem tedious but an unnecessary code replication: Wouldn't it be more reasonable to allow providers (incl. the test `tls-provider') to also register "default supported groups"? Issue #13831 has been created to formally propose this.

Question now: Does #13831 stand a chance of getting accepted? If so, I'll work on this, otherwise I'll complete a PR changing test cases such as above to use (test-) provider groups in the absence of (EC+DH) groups to close this issue.

**Comment by mattcaswell** [comment]
> Update after some code changes: It seems possible to break the (now outdated) link "no-ec&&no-dh -> no tls1_3" by adding || !defined(OPENSSL_NO_TLS1_3) to places like

The "no-ec"/"no-dh" -> "no-tls1_3" link is fixed by #13916
